keps/sig-network/2136-np-version-status/README.md

```diff
@@ -0,0 +1,789 @@
+    "Supported", status: "Unknown", reason: "NoMinVersion" }`. (In the
+    absence of `minVersion`, it is impossible for the plugin to know for
+    sure that the policy does not use some new feature that the plugin
+    is unaware of.)
+
+Note the distinction between the first two cases; if the `minVersion` is
+higher than the highest version the plugin knows about, then the plugin
+has no way of telling whether it can implement the policy or not, and no
+way of knowing for sure if it's correctly interpreting even the parts
+that it can see. (For example, if a network plugin that only implemented
+version "1.9" received a policy with a combined `podSelector` and
+`namespaceSelector`, it might misinterpret it by reading the
+`namespaceSelector` but not the `podSelector`, and thereby allowing
+access to/from more pods than was intended.) Thus, it is best for it to
+not even try.
+
+On the other hand, if the plugin knows about a feature but doesn't
+implement it, they may still be able to implement the other parts of the
+policy safely. (For example, a plugin that does not implement egress
+policies at all might choose to still implement the ingress side of a
+mixed ingress/egress policy.)
+
+#### The "Enforcing" Condition
+
+The `"Enforcing"` condition indicates whether the plugin is currently
+enforcing the policy described by a NetworkPolicy.
+
+- If the plugin is intentionally not enforcing the policy (eg, because
+  it depends on unsupported features), it should set a `{ type: "Enforcing",
+  status: "False" }` condition, with an appropriate `message` and the
+  same `reason` as the `"Supported"` condition.
+
+- A plugin which is implementing a policy but is not able to determine
+  when the policy is fully in effect should set a status condition with
+  `{ type: "Enforcing", status: "Unknown" }` and an appropriate `reason`
+  and `message`. This indicates to the user that the enforcing status is
+  not known and is not going to become known.
+
+- A plugin which is able to determine when a NetworkPolicy is fully in
+  effect _may_ set an initial status condition with `{ type:
+  "Enforcing", status: "False", reason: "Pending" }`, and an appropriate
+  `message`. In particular, it should do this if it thinks there may be
+  a noticeable delay before it is able to set the condition to `"True"`.
+
+- A plugin which is able to determine when a NetworkPolicy is fully in
+  effect should set a status condition with `{ type: "Enforcing",
+  status: "True" }`, and an appropriate `reason` and `message`, once it
+  knows the NetworkPolicy is fully in effect.
+
+  "Fully in effect" means that for any pod which ought to be subject
+  to the policy, either (a) the pod was subject to the policy at the time
+  when the `"Enforcing"` condition became `"True"`, or (b) the pod was not
+  yet "Ready" at the time the `"Enforcing"` condition became `"True"`,
+  but will not become "Ready" until it is subject to the policy.
+
+```
+<<[UNRESOLVED pod-readiness ]>>
```

Comment:
hmmm...  i wonder if maybe the readiness thing is really required here.  it might add semantics that can confuse people.  enforcing is clear - its a CNI providers way of saying "to the best of my knowledge, ive implemented the rules for this policy" 

Readiness is more an eye of the beholder thing, which is really tough to reason about (loopback use case again), or maybe some masquerading use cases might trip that up ?

Reply:
But what does "enforcing" mean without "readiness"?

If I create a Pod and a NetworkPolicy at about the same time, and the Pod becomes Ready and the NetworkPolicy becomes Enforcing, then is it guaranteed that the Pod is subject to the NetworkPolicy? If you say "no", then "Enforcing" is much less useful. If you say "yes" then that implies a mandatory connection between NP Enforcing and Pod Readiness.

I think people generally _want_ Pod Readiness to track NP enforcement when possible...

Reply:
I guess this hinges on what we mean by being "subject" to a policy:
- am i subject to a policy `iff` the policy is *correctly* implemented and no unauthorized entities can commmuncate with me  (call it NUE for abbrev).
OR
- am i subject to a policy `iff` my CNI provider , to the best of its abilities, has tried to update its firewall/iptables/ovs/whatever rules to enforce "NUE".

Reply:
The former. If `Enforcing True` only means "I'm _pretty sure_ that the rules are working now", then the e2e suite can't trust it, and so it still needs to just retry for some arbitrary length of time after that, and if it's going to do that, there's really not much point in having the status at all.

For iptables, once the rules are in the kernel, they're in effect. So if you know you have called `iptables` or `iptables-restore` on each relevant node, and gotten back success, then you know NUE. If you've merely set a flag indicating that you plan to update iptables the next time a certain timer goes off, then you don't claim `Enforcing` yet.

Reply:
> once the rules are in the kernel they're in effect
what if theres a bug in iptables or a CVE or something?

Reply:
If there's a bug then buggy behavior will occur. That goes without saying. The spec is defining what correct behavior is, not magically obligating you to deliver only correct code.

In this particular case, if there was some bug in iptables that caused a plugin to report `Enforcing True` when it wasn't actually enforcing, then that would cause the e2e tests to flake, and this would be noticed, and the plugin author would need to decide whether to (a) remain flaky, (b) start reporting `{ type: Enforcing, status: Unknown, reason: IPTables, message: "Unable to determine enforcing status due to iptables bugs" }` instead, (c) find a workaround for the iptables bug, (d) debug and fix the iptables bug.

Reply:
acknowledged..  worth clarifying in this KEP imo , but your argument is sound

Reply:
I may be going far from the original Kep proposal here, but maybe this concern here can be solved with something like  a Pod that can verify if the current Enforcing status reflects what is expected (like 'API is reporting status.Enforcing = true so I can try to connect to whatever rule has been created), and if it can't, exit with 1, leading the Pod to an error that can be reported to the e2e test.

This approach seems reasonable to me, so you can have a 'common network policy tester image' that can wait for the condition, test and check if the plugin is reporting correctly